onnx/ir.h

```diff
@@ -0,0 +1,1048 @@
+  // Given:   %3 = f(%1, %2)
+  //          %4 = g(%3)
+  //          %5 = h(%3, %3)
+  // Execute: %3.replaceAllUsesWith(%6)
+  // Result:  %3 = f(%1, %2)
+  //          %4 = g(%6)
+  //          %5 = h(%6, %6)
+  void replaceAllUsesWith(Value * newValue);
+
+  Value* copyMetadata(Value * from) {
+    setElemType(from->elemType());
+    setSizes(from->sizes());
+    if (!from->unique_name_.empty()) {
+      setUniqueName(from->uniqueName());
+    }
+    return this;
+  }
+
+};
+
+
+struct Node : public Attributes<Node> {
+  ONNX_DISALLOW_COPY_AND_ASSIGN(Node);
+  friend struct Graph;
+  friend struct Value;
+  friend graph_node_list;
+  friend const_graph_node_list;
+  friend graph_node_list_iterator;
+  friend const_graph_node_list_iterator;
+
+private:
+  // each node but Return/Param
+  // is associated with exactly one place in the node list...
+  // of the graph_
+  // this circular is a doubly-linked list, the Return node is used as the sentinel for the beginning and end of the list
+  // such that the list never has null pointers
+  // next_in_graph[0] is next pointer
+  // next_in_graph[1] is prev pointer
+  // using an array to allow the same iterator class for forward and reverse node lists
+  // This list represents a topological sort
+
+  Node* next_in_graph[2] = { nullptr, nullptr };
+  Node* & next() { return next_in_graph[kNextDirection]; }
+  Node* & prev() { return next_in_graph[kPrevDirection]; }
+  Node* const & next() const { return next_in_graph[kNextDirection]; }
+  Node* const & prev() const { return next_in_graph[kPrevDirection]; }
+
+  const NodeKind kind_;
+  std::vector<Value*> inputs_;
+  std::vector<Value*> outputs_;
+  Graph* graph_;
+  size_t stage_;
+  bool has_name_;
+  std::string name_;
+  bool has_doc_string_;
+  std::string doc_string_;
+
+protected:
+  Node(Graph * graph_, NodeKind kind_); //defined after graph
+
+public:
+  bool has_name() {
+    return has_name_;
+  }
+  const std::string& name() {
+    return name_;
+  }
+  void setName(const std::string& name) {
+    has_name_ = true;
+    name_ = name;
+  }
+  bool has_doc_string() {
+    return has_doc_string_;
+  }
+  const std::string& docString() {
+    return doc_string_;
+  }
+  void setDocString(const std::string& doc_string) {
+    has_doc_string_ = true;
+    doc_string_ = doc_string;
+  }
+  NodeKind kind() const {
+    return kind_;
+  }
+  Graph * owningGraph() {
+    return graph_;
+  }
+  const Graph * owningGraph() const {
+    return graph_;
+  }
+  size_t stage() const {
+    return stage_;
+  }
+  Node* setStage(size_t s) {
+    stage_ = s;
+    return this;
+  }
+  // NB: This returns an ArrayRef; that means that it will
+  // get invalidated if you resize inputs (e.g., using addInput)
+  // We can't return a std::vector<Node*>& because there's no
+  // way to soundly cast to std::vector<const Node*> (an insane
+  // implementation of std::vector could make this representationally
+  // different.)
+  ArrayRef<Value*> inputs() {
```

Comment:
I guess... that would work. But you'd have to maintain a canonical `inputs_array_ref_` in the struct (because the reference needs to point to something) and then remember to update it whenever you update `std::vector`, not to mention how easy it would be to accidentally make a copy of `ArrayRef` (because it is intended to be a pass-by-value type). Seems not worth it.